### DISCUSSION TITLE
WinUI 3 ItemsRepeater scrolling is slow

### DISCUSSION THREAD

**Discussion by hshristov** [microsoft/microsoft-ui-xaml#7514] [Q&A]
When scrolling an **ItemsRepeater** with 10 buttons (in horizontal StackPanel) as an **ItemTemplate** the frame per seconds are below 10. Profiling shows that most of the time is spent in **Layout** (e.g. measure & arrange).
Is there a way to reach at least 30 fps or even better 60 fps?

Check the attached project for reference. You will also notice that when dragging the scrollbar thumb there is a time where a white screen is shown and also thumb position is not always under the mouse cursor.
[RepeaterSlowScroll.zip](https://github.com/microsoft/microsoft-ui-xaml/files/4994777/RepeaterSlowScroll.zip)

**Comment by StephenLPeters** [comment]
@jeffStall We've had some reported scrolling issues that I think ended up on IXP, could this be a dupe? From the description it sounds like it is Dmanip (from scrollviewer) though.  @ranjeshj has seen issues with EffectiveViewport which may also be contributing here.

**Comment by ranjeshj** [comment]
@hshristov Couple of things might be contributing to slower scrolling performance. 10 buttons in the item template expand into several elements in each item. Depending on how big your window is you get a page full of items + 4 more pages realized in order to ensure that you don't see empty space when you scroll. That could amount to a lot of elements that are going through layout when you start scrolling. Some thoughts to improve the perf
* Try reducing the number of elements in the item template (look at live visual tree to see the counts). Each element element you save gets multiplied due to the repeating nature of this scenario.
* Play with the VerticalCacheLength property to reduce the 4 pages of buffer that I mentioned earlier. 

I have not analyzed perf on your app, so there could possibly be other things at play.

The height of each item is not uniform, and stack layout is a virtualizing layout. It does not measure everything to accurately know what the real extent is. It is estimating the extent as it sees more items. That is the reason for the change in scroll bar size and it not sticking to the mouse.
